Firestore/core/src/firebase/firestore/remote/remote_objc_bridge.mm

```diff
@@ -127,170 +112,233 @@ bool IsLoggingEnabled() {
+WriteStreamSerializer::WriteStreamSerializer(Serializer serializer)
+    : serializer_{std::move(serializer)} {
 }
 
-GCFSWriteRequest* WriteStreamSerializer::CreateHandshake() const {
-  // The initial request cannot contain mutations, but must contain a projectID.
-  GCFSWriteRequest* request = [GCFSWriteRequest message];
-  request.database = [serializer_ encodedDatabaseID];
-  return request;
+Message<google_firestore_v1_WriteRequest>
+WriteStreamSerializer::CreateHandshake() const {
+  Message<google_firestore_v1_WriteRequest> result{
+      google_firestore_v1_WriteRequest_fields};
+  auto& request = result.mutable_proto();
+
+  // The initial request cannot contain mutations, but must contain a project
+  // ID.
+  request.database = serializer_.EncodeDatabaseId();
+
+  return result;
 }
 
-GCFSWriteRequest* WriteStreamSerializer::CreateWriteMutationsRequest(
-    const std::vector<Mutation>& mutations) const {
-  NSMutableArray<GCFSWrite*>* protos =
-      [NSMutableArray arrayWithCapacity:mutations.size()];
-  for (const Mutation& mutation : mutations) {
-    [protos addObject:[serializer_ encodedMutation:mutation]];
-  };
+Message<google_firestore_v1_WriteRequest>
+WriteStreamSerializer::CreateWriteMutationsRequest(
+    const std::vector<Mutation>& mutations,
+    const ByteString& last_stream_token) const {
+  Message<google_firestore_v1_WriteRequest> result{
+      google_firestore_v1_WriteRequest_fields};
+  auto& request = result.mutable_proto();
 
-  GCFSWriteRequest* request = [GCFSWriteRequest message];
-  request.writesArray = protos;
-  request.streamToken = MakeNullableNSData(last_stream_token_);
+  if (!mutations.empty()) {
+    request.writes_count = nanopb::CheckedSize(mutations.size());
+    request.writes = MakeArray<google_firestore_v1_Write>(request.writes_count);
 
-  return request;
-}
+    for (pb_size_t i = 0; i != request.writes_count; ++i) {
+      request.writes[i] = serializer_.EncodeMutation(mutations[i]);
+    }
+  }
+
+  request.stream_token = nanopb::CopyBytesArray(last_stream_token.get());
 
-grpc::ByteBuffer WriteStreamSerializer::ToByteBuffer(
-    GCFSWriteRequest* request) {
-  return ConvertToByteBuffer([request data]);
+  return result;
 }
 
-GCFSWriteResponse* WriteStreamSerializer::ParseResponse(
-    const grpc::ByteBuffer& message, Status* out_status) const {
-  return ToProto<GCFSWriteResponse>(message, out_status);
+MaybeMessage<google_firestore_v1_WriteResponse>
+WriteStreamSerializer::ParseResponse(const grpc::ByteBuffer& message) const {
+  return Message<google_firestore_v1_WriteResponse>::Parse(
+      google_firestore_v1_WriteResponse_fields, message);
 }
 
 model::SnapshotVersion WriteStreamSerializer::ToCommitVersion(
-    GCFSWriteResponse* proto) const {
-  return [serializer_ decodedVersion:proto.commitTime];
+    const google_firestore_v1_WriteResponse& proto) const {
+  nanopb::Reader reader;
+  auto result = serializer_.DecodeSnapshotVersion(&reader, proto.commit_time);
```

Comment:
Theoretically, `reader` may have an error, though the function signature currently has no way to report that. Do you have any thoughts on how to approach error handling here?

Reply:
Ultimately we need to decide if we're going to propagate this error out to somewhere such that this response can be handled as broken or we have to assert.

To do the former it seems like this has to return `StatusOr<SnapshotVersion>`, no?

If we're going to do that everywhere I suggest a method like the following on reader:

```c++
template <typename T>
StatusOr<T> ToStatusOr(T&& value) {
  if (ok()) {
    return std::move(value);
  } else {
    return status();
  }
}
```

Reply:
Let's discuss this. The only two alternatives I can think of are either asserting that reader is `ok`, or else returning `StatusOr` pretty much everywhere. I presume the former is closer to what we had before (Objective-C proto library would probably throw upon failure). It sounds like you would prefer the latter, using `StatusOr` throughout?